### ISSUE TITLE
Possible bug / Documentation Suggestion - Dataset API + Estimator API does not throw an exception when the input filepath to the dataset is incorrect

### ISSUE THREAD

**Issue by tejaswid** [tensorflow/tensorflow#21335]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: v1.9.0-0-g25c197e023 1.9.0
- **Python version**: 3.6.6
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: 
- **CUDA/cuDNN version**: CUDA 9.0  / CUDNN 7.1.4.18
- **GPU model and memory**: GTX 1080 Ti - 9710 MB
- **Exact command to reproduce**: See below

### Describe the problem
When using the Dataset API along with the Estimator API, if the input file path to the dataset, while initializing **tf.data.TFRecordDataset** or **tf.data.FixedLengthRecordDataset**, is incorrect/ file does not exist, it does not immediately throw an exception. Instead the entire network graph is constructed, weights are initialized and checkpoints (if present) are loaded.

The only indication that something is wrong is when the session.run is called inside the function **_train_with_estimator_spec** in the train loop of the estimator.** The loss remains _None_ and the global_step is not updated**. Moreover, the training runs for as many epochs as requested with nothing being updated. 

However, even if one were to use the Dataset API with an incorrect file path but still overwrite the features and labels returned by it with manually set values, the loss is still not computed and remains as _None_ and the same problem as above persists. This is very counter-intuitive and hard to debug.

It would be better if the Dataset API somehow threw an exception in the beginning itself. If this is not possible, it would be nice if this was documented somewhere.

### Source code

1. Clone the tensorflow models from the official repository.
`git clone https://github.com/tensorflow/models.git tfmodels`
`cd tfmodels`

2. Run the **cifar10_download_and_extract.py** file to download and extract the cifar10 dataset.
3. Run the **cifar10_main.py** file to check that the original model is running correctly.

4. Delete the .bin files inside the **cifar-10-batches-bin** folder that was created in step 2. Keep the folder but only delete the contents inside it. By doing so we are passing an empty string to the dataset initializer and hence the dataset iterator is probably returning garbage values.
5. Re-run step 3 and you see the error described.

6. Make a copy of the file **cifar10_main.py** and edit it by adding the following lines just before the return statement in the function **parse_record**.
`image = tf.constant(value=0.0, shape=[32, 32, 3], dtype=tf.float32)`
`label = tf.constant(value=0, shape=[], dtype=tf.int32)`
By doing so we are still using an incorrect initializer but are now replacing the garbage values returned by the reader with known constants (zeros in this case).
7. Re-run the modified file and you still see the error described.

**Comment by karmel** [comment]
Thanks, @tejaswid , that does seem less than ideal. @mrry , can you comment on the expected behavior of Dataset if a passed path is incorrect, separate from estimator? Do we currently do any checking there, or is it assumed that is the role of clients?

**Comment by mrry** [comment]
I think this is working as intended: the iterator should raise an exception (I think it would be `tf.errors.InvalidArgumentError`) when the first element is retrieved. 

However, I'm surprised by this part of the bug report:

> By doing so we are passing an empty string to the dataset initializer and hence the dataset iterator is probably returning garbage values.

Can you try reproducing this problem with a simpler example? As mentioned above, the iterator should raise an exception and never generate garbage values when the path is unrecognized.

**Comment by tejaswid** [comment]
When I run the following simple example,
```python
import tensorflow as tf

def _parse_data(raw_tfrecord):
    feature = {'/value': tf.FixedLenFeature([], tf.int64)}
    data = tf.parse_single_example(raw_tfrecord, features=feature)
    value = tf.cast(data['/value'], tf.int32)
    return value

data = tf.data.TFRecordDataset("<FILENAME>")  # see below for explanation
data_parsed = data.map(_parse_data)
batched_dataset = data_parsed.batch(1)
iterator = batched_dataset.make_initializable_iterator()
next_element = iterator.get_next()

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    sess.run(iterator.initializer)
    print(sess.run(next_element))
    sess.close()
```
If `<FILENAME>` is a non-existent file (like empty string) the error says
```
NotFoundError (see above for traceback): ; No such file or directory
[[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?]], output_types=[DT_STRING], _device="/job:localhost/replica:0/task:0/device:CPU:0"](Iterator)]]
```
If `<FILENAME>` is an existing directory but not a _.tfrecords_ file the error says
```
FailedPreconditionError (see above for traceback): /media/drzadmin/Z/repos/python/datasets; Is a directory
	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?]], output_types=[DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](Iterator)]]
```
which I think is the expected behaviour.

When I run the code mentioned in my first post, no error is thrown. Instead loss is not computed at all and remains **None**. The following lines in the function **_train_with_estimator_spec** in **estimator.py** is where the loss should actually be computed, but it remains None after the sess.run call. Furthermore the training continues but the **global_step value always remains at 0**.

```python
loss = None
while not mon_sess.should_stop():
    _, loss = mon_sess.run([estimator_spec.train_op, estimator_spec.loss])
return loss
```

**Comment by facaiy** [comment]
I can reproduce the bug with [models/official/resnet/cifar10_main.py](https://github.com/tensorflow/models/blob/master/official/resnet/cifar10_main.py) . And the exception of [_WrappedSession](https://github.com/tensorflow/tensorflow/blob/aed8f42bafabf11c5d92ce4109a5e0408b31f9c5/tensorflow/python/training/monitored_session.py#L987) is shown below when running:

```python
End of sequence
         [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,32,32,3], [?]], output_types=[DT_FLOAT, DT_INT32], _device="/device:CPU:0"](IteratorFromStringHandle)]]
         [[Node: FunctionBufferingResourceGetNext = FunctionBufferingResourceGetNext[output_types=[DT_FLOAT, DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](FunctionBufferingResource)]]
coord exception: End of sequence
         [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,32,32,3], [?]], output_types=[DT_FLOAT, DT_INT32], _device="/device:CPU:0"](IteratorFromStringHandle)]]
         [[Node: FunctionBufferingResourceGetNext = FunctionBufferingResourceGetNext[output_types=[DT_FLOAT, DT_INT32], _device="/job:localhost/replica:0/task:0/device:CPU:0"](FunctionBufferingResource)]]
```

**Comment by karmel** [comment]
@robieta - can you investigate whether this is an issue with Cifar10 in particular, or more generally?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @robieta: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @robieta: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @karmel, @robieta: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sachinprasadhs** [comment]
EstimatorSpec will be deprecated from Tensorflow 2.12 version, https://www.tensorflow.org/api_docs/python/tf/estimator/EstimatorSpec?version=nightly.
Could you please migrate from Estimator to Keras API which will enable you to perform the aforementioned tasks, such as [model building](https://www.tensorflow.org/guide/keras/custom_layers_and_models), gradient application, [training](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit), evaluation, and prediction.
Refer Migration guide here https://www.tensorflow.org/guide/migrate/migrating_estimator for more details. Thanks!

**Comment by tejaswid** [comment]
Sure. I am closing this thread here as this code will now be deprecated.
